lib/maths/CTreeShapFeatureImportance.cc

```diff
@@ -0,0 +1,258 @@
+}
+
+CTreeShapFeatureImportance::TDoubleVec
+CTreeShapFeatureImportance::samplesPerNode(const TTree& tree,
+                                           const core::CDataFrame& frame,
+                                           const CDataFrameCategoryEncoder& encoder,
+                                           std::size_t numThreads) {
+    auto result = frame.readRows(
+        numThreads, 0, frame.numberRows(),
+        core::bindRetrievableState(
+            [&](TDoubleVec& state, TRowItr beginRows, TRowItr endRows) {
+                for (auto row = beginRows; row != endRows; ++row) {
+                    auto encodedRow{encoder.encode(*row)};
+                    auto node{tree[0]};
+                    state[0] += 1;
+                    std::size_t nextIndex;
+                    while (node.isLeaf() == false) {
+                        if (node.assignToLeft(encodedRow)) {
+                            nextIndex = node.leftChildIndex();
+                        } else {
+                            nextIndex = node.rightChildIndex();
+                        }
+                        state[nextIndex] += 1.0;
+                        node = tree[nextIndex];
+                    }
+                }
+            },
+            TDoubleVec(tree.size())));
+
+    auto& state = result.first;
+    TDoubleVec totalSamplesPerNode{std::move(state[0].s_FunctionState)};
+    for (std::size_t i = 1; i < state.size(); ++i) {
+        for (std::size_t nodeIndex = 0; nodeIndex < totalSamplesPerNode.size(); ++nodeIndex) {
+            totalSamplesPerNode[nodeIndex] += state[i].s_FunctionState[nodeIndex];
+        }
+    }
+
+    return totalSamplesPerNode;
+}
+
+CTreeShapFeatureImportance::CTreeShapFeatureImportance(TTreeVec trees, std::size_t threads)
+    : m_Trees{std::move(trees)}, m_NumberThreads{threads}, m_SamplesPerNode() {
+    m_SamplesPerNode.reserve(m_Trees.size());
+}
+
+std::size_t CTreeShapFeatureImportance::updateNodeValues(TTree& tree,
+                                                         std::size_t nodeIndex,
+                                                         const TDoubleVec& samplesPerNode,
+                                                         std::size_t depth) {
+    auto& node{tree[nodeIndex]};
+    if (node.isLeaf()) {
+        return 0;
+    }
+
+    std::size_t depthLeft = CTreeShapFeatureImportance::updateNodeValues(
+        tree, node.leftChildIndex(), samplesPerNode, depth + 1);
+    std::size_t depthRight = CTreeShapFeatureImportance::updateNodeValues(
+        tree, node.rightChildIndex(), samplesPerNode, depth + 1);
+
+    double leftWeight = samplesPerNode[node.leftChildIndex()];
+    double rightWeight = samplesPerNode[node.rightChildIndex()];
+    double averageValue = (leftWeight * tree[node.leftChildIndex()].value() +
+                           rightWeight * tree[node.rightChildIndex()].value()) /
+                          (leftWeight + rightWeight);
+    node.value(averageValue);
```

Comment:
👍 I like the idea of repurposing the internal node values for this! It is useful information to surface in its own right.